Create rule S6838: @Bean methods for Singleton should not be invoked in @Configuration when proxyBeanMethods is false (#3367)
SonarSource/rspec: rules/S6838/java/rule.adoc

Proxies in Spring are an implementation of the Proxy design pattern that allows to create an intermediary that acts as an interface to another resource.
The proxy provides additional functionality to the component it represents by paying a small performance penalty.
For example, it enables lazy initialization of the resource, or caching of data.

> The `@Configuration` sets the `proxyBeanMethods` attribute to `true` by default.

Reviewer: ```suggestion
The `@Configuration` annotation enables this mechanism by setting the `proxyBeanMethods` attribute to `true` by default.
```

This means that the `@Bean` methods are proxied in order to enforce bean lifecycle behavior, e.g. to return shared singleton bean instances even in case of direct `@Bean` method calls in user code.

> This feature requires method interception, implemented through a runtime-generated *CGLIB* subclass.

Reviewer: Should we add a link to document CGLIB or remove the last part of the sentence altogether?
```suggestion
This feature requires method interception.
```

Reviewer 2: I'll add a link. In my opinion is valuable to reference it, since the main reason for disabling the `proxyBeanMethods` is to avoid the overhead  due to CGLIB, and it only takes a few words to mention it

== Why is this an issue?

When setting the `proxyBeanMethods` attribute to `false` the `@Bean` methods are not proxied.
This is behaviorally equivalent to removing the `@Configuration` stereotype.
In these cases the `@Bean` methods are processed in https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/annotation/Bean.html[_lite mode_], meaning that the bean is created every time the method is called.

> In case of `Singleton` beans, this can lead to unexpected behavior, as the bean is created multiple times, instead of being created once and cached.

Reviewer: Are we talking about beans that are effectively singletons or beans annotated with `@Singleton`?

Reviewer: After discussion: Maybe we could repeat here that bean methods return the same instance of an object by default unless marked otherwise.
```suggestion
This changes from the default behavior of `@Bean` methods returning the same instance unless the method is instructed otherwise (eg: when the type returned by the method is annotated as `@Prototype`).
In case of `Singleton` beans, this can lead to unexpected behavior, as the bean is created multiple times, instead of being created once and cached.
```

Reviewer 2: By default, all beans are singleton in Spring. So it refers to beans intended to be singletons

Reviewer: I still think it is worth mentioning the effect of `@Scope("prototype")` either in the example itself or in the text above

The rule raises an issue when the `proxyBeanMethods` attribute is set to `false` and the `@Bean` method is called directly in the code for a `Singelton` bean.

== How to fix it

> The rule can be fixed in the following ways:

Reviewer: ```suggestion
The issue can be fixed in the following ways:
```

* Not invoking the `@Bean` method directly, but rather injecting the bean in the context and using it, for example, by auto-wiring it.

* If the performance penalty is negligible, consider not disabling the `proxyBeanMethods` attribute, so that the `@Bean` methods are proxied and the bean lifecycle is enforced.

=== Code examples

==== Noncompliant code example

In the example below, every instance of `PrototypeBean` will have a different instance of `SingletonBean`, as the `@Bean` method is called directly in the code.

[source,java,diff-id=1,diff-type=noncompliant]
----
@Configuration(proxyBeanMethods = false)
class ConfigurationExample {
  @Bean
  public SingletonBean singletonBean() {
    return new SingletonBean();
  }

  @Bean
  @Scope("prototype")
  public PrototypeBean prototypeBean() {
    return new PrototypeBean(singletonBean()); // Noncompliant, the bean is created every time the method is called
  }

  class SingletonBean {
    // ...
  }

  class PrototypeBean {
    // ...

    public PrototypeBean(SingletonBean singletonBean) {
      // ...
    }

    // ...
  }
}
----

==== Compliant solution

> The compliant solution consists in injecting the `SingletonBean` in the context and using it in the `PrototypeBean` constructor. This way every instance of `PrototypeBean` will have the same instance of `SingletonBean`.

Reviewer: Let's break these two sentences apart onto their own respective lines
```suggestion
The compliant solution consists in injecting the `SingletonBean` in the context and using it in the `PrototypeBean` constructor.
This way every instance of `PrototypeBean` will have the same instance of `SingletonBean`.
```

[source,java,diff-id=1,diff-type=compliant]
----
@Configuration(proxyBeanMethods = false)
class ConfigurationExample {
  @Bean
  public SingletonBean singletonBean() {
    return new SingletonBean();
  }

  @Bean
  @Scope("prototype")

>   public PrototypeBean prototypeBean(SingletonBean singletonBean) { // Compliant, the bean is injected in the context and used

Reviewer: Because proxying is off, isn't it always going to be a different object?

Reviewer 2: When the @Bean method has arguments, these are injected by the framework from the ApplicationContext. 
Maybe it is a good idea to reference this link: https://docs.spring.io/spring-framework/reference/core/beans/java/bean-annotation.html#beans-java-dependencies

Reviewer: Thanks for clarifying
